src/webpush/mod.rs

```diff
@@ -0,0 +1,179 @@
+            #[derive(RustcDecodable, Debug)]
+            struct SubscribeBody {
+                user: Option<String>,
+                push_uri: String,
+                push_key: String,
+                resources: Vec<String>
+            }
+
+            let mut payload = String::new();
+            req.body.read_to_string(&mut payload).unwrap();
+            let body: SubscribeBody = match json::decode(&payload) {
+                Ok(body) => body,
+                Err(e) => {
+                    warn!("invalid subscribe payload {:?}", e);
+                    return Ok(Response::with(status::BadRequest));
+                }
+            };
+
+            let db = ctl.get_users_manager().get_db();
+            let mut user = if cfg!(feature = "authentication") {
+                req.extensions.get::<User>().unwrap().clone()
+            } else {
+                let name = match body.user {
+                    Some(n) => n,
+                    None => {
+                        warn!("not using auth and no user provided");
+                        return Ok(Response::with(status::BadRequest));
+                    }
+                };
+
+                let mut users = match db.read(ReadFilter::Name(name)) {
+                    Ok(u) => u,
+                    Err(_) => Vec::new()
+                };
+
+                if users.len() != 1 {
+                    warn!("user not found in database");
+                    return Ok(Response::with(status::BadRequest));
+                }
+
+                users.pop().unwrap().clone()
+            };
+
+            user.push_uri = if body.push_uri.is_empty() {
+                None
+            } else {
+                Some(body.push_uri)
+            };
+            user.push_key = if body.push_key.is_empty() {
+                None
+            } else {
+                Some(body.push_key)
+            };
+
+            let id = user.id.unwrap();
+            if let Err(e) = db.update(id, &user) {
+                warn!("cannot update push subscription for user {}: {:?}", user.name, e);
+                return Ok(Response::with(status::InternalServerError));
+            }
+            if let Err(e) = db.update_push_resources(id, &body.resources) {
+                warn!("cannot update push resources for user {}: {:?}", user.name, e);
+                return Ok(Response::with(status::InternalServerError));
+            }
+
+            Ok(Response::with(status::Ok))
+        });
+
+        /*
+        let wp = controller.get_web_push();
+        router.post("test", move |_: &mut Request| -> IronResult<Response> {
+            wp.notify(String::from("mygroup"), String::from("mymessage"));
+            Ok(Response::with(status::Ok))
+        });*/
+
+        let auth_endpoints = if cfg!(feature = "authentication") {
+            vec![
+                AuthEndpoint(vec![Method::Post], "subscribe".to_owned())
+            ]
+        } else {
+            vec![]
+        };
```

Comment:
I just realized here that I did not provide any way for the user to query their current subscriptions. I will add that.